db/sql_implementation.go

```diff
@@ -452,40 +484,35 @@ func (db *DB) UpdateOrder(hash common.Hash, updateFunc func(existingOrder *types
 		return errors.New("db.UpdateOrders: updateFunc cannot be nil")
 	}
 
-	txn, err := db.sqldb.BeginTxx(db.ctx, nil)
-	if err != nil {
-		return err
-	}
-	defer func() {
-		_ = txn.Rollback()
-	}()
-
-	var existingOrder sqltypes.Order
-	if err := txn.GetContext(db.ctx, &existingOrder, "SELECT * FROM orders WHERE hash = $1", hash); err != nil {
-		if err == sql.ErrNoRows {
-			return ErrNotFound
+	return db.sqldb.TransactionalContext(db.ctx, nil, func(txn *sqlz.Tx) error {
+		var existingOrder sqltypes.Order
+		if err := txn.GetContext(db.ctx, &existingOrder, "SELECT * FROM orders WHERE hash = $1", hash); err != nil {
+			if err == sql.ErrNoRows {
+				return ErrNotFound
+			}
+			return err
 		}
-		return err
-	}
 
-	commonOrder := sqltypes.OrderToCommonType(&existingOrder)
-	commonUpdatedOrder, err := updateFunc(commonOrder)
-	if err != nil {
-		return fmt.Errorf("db.UpdateOrders: updateFunc returned error")
-	}
-	updatedOrder := sqltypes.OrderFromCommonType(commonUpdatedOrder)
-	_, err = txn.NamedExecContext(db.ctx, updateOrderQuery, updatedOrder)
-	if err != nil {
-		return err
-	}
-	return txn.Commit()
+		commonOrder := sqltypes.OrderToCommonType(&existingOrder)
+		commonUpdatedOrder, err := updateFunc(commonOrder)
+		if err != nil {
+			return fmt.Errorf("db.UpdateOrders: updateFunc returned error")
+		}
+		updatedOrder := sqltypes.OrderFromCommonType(commonUpdatedOrder)
+		_, err = txn.NamedExecContext(db.ctx, updateOrderQuery, updatedOrder)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
 }
 
 func (db *DB) AddMiniHeaders(miniHeaders []*types.MiniHeader) (added []*types.MiniHeader, removed []*types.MiniHeader, err error) {
 	defer func() {
 		err = convertErr(err)
 	}()
-	var miniHeadersToRemove []*sqltypes.MiniHeader
+
+	addedMap := map[common.Hash]*types.MiniHeader{}
```

Comment:
I found a more efficient way to remove duplicates from the `added` and `removed` sets. Applied the same approach to both `AddMiniHeaders` and `AddOrders` for both the SQL and Dexie implementations.